s3/pkg/service/multipart.go

```diff
@@ -0,0 +1,451 @@
+		log.Errorln("failed to receive msg. err:", err)
+		return err
+	}
+	bucketName := uploadRequest.BucketName
+	objectKey := uploadRequest.ObjectKey
+	partId := uploadRequest.PartId
+	uploadId := uploadRequest.UploadId
+	size := uploadRequest.Size
+
+	log.Infof("receive msg, bucketname:%v, objectkey:%v, partId:%v, uploadId:%v,size:%v", bucketName, objectKey, partId, uploadId, size)
+	multipart, err := s.MetaStorage.GetMultipart(bucketName, objectKey, uploadId)
+	if err != nil {
+		log.Infoln("failed to get multipart. err:", err)
+		return err
+	}
+
+	if size > MAX_PART_SIZE {
+		log.Errorf("object part size is too large. size:", size)
+		err = ErrEntityTooLarge
+		return err
+	}
+
+	isAdmin, tenantId, err := util.GetCredentialFromCtx(ctx)
+	if err != nil && isAdmin == false {
+		log.Error("get tenant id failed")
+		err = ErrInternalError
+		return err
+	}
+
+	bucket, err := s.MetaStorage.GetBucket(ctx, bucketName, true)
+	if err != nil {
+		log.Errorln("get bucket failed with err:", err)
+		return err
+	}
+
+	if !isAdmin {
+		switch bucket.Acl.CannedAcl {
+		case "public-read-write":
+			break
+		default:
+			if bucket.TenantId != tenantId {
+				err = ErrBucketAccessForbidden
+				log.Errorln("bucket owner is not equal to request owner.")
+				return err
+			}
+		}
+	}
+
+	backendName := bucket.DefaultLocation
+	backend, err := utils.GetBackend(ctx, s.backendClient, backendName)
+	if err != nil {
+		log.Errorln("failed to get backend client with err:", err)
+		return err
+	}
+
+	data := &StreamReader{in: stream}
+	limitedDataReader := io.LimitReader(data, size)
+	sd, err := driver.CreateStorageDriver(backend.Type, backend)
+	if err != nil {
+		log.Errorln("failed to create storage. err:", err)
+		return nil
+	}
+	ctx = context.WithValue(ctx, dscommon.CONTEXT_KEY_MD5, uploadRequest.Md5Hex)
+	log.Infoln("bucketname:", bucketName, " objectKey:", objectKey, " uploadid:", uploadId, " objectId:", multipart.ObjectId, " partid:", partId)
+	result, err := sd.UploadPart(ctx, limitedDataReader, &pb.MultipartUpload{
+		Bucket:   bucketName,
+		Key:      objectKey,
+		UploadId: uploadId,
+		ObjectId: multipart.ObjectId},
+		int64(partId), size)
+	if err != nil {
+		log.Errorln("failed to upload part to backend. err:", err)
+		return err
+	}
+	uploadResponse.ETag = result.ETag
+
+	log.Infoln("UploadPart upload part successfully.")
+	return nil
+}
+
+func (s *s3Service) CompleteMultipartUpload(ctx context.Context, in *pb.CompleteMultipartRequest, out *pb.CompleteMultipartResponse) error {
+	log.Info("CompleteMultipartUpload is called in s3 service.")
+	bucketName := in.BucketName
+	objectKey := in.ObjectKey
+	uploadId := in.UploadId
+
+	var err error
+	completeMultipartResponse := pb.BaseResponse{}
+	defer func() {
+		completeMultipartResponse.ErrorCode = GetErrCode(err)
+	}()
+
+	isAdmin, tenantId, err := util.GetCredentialFromCtx(ctx)
+	if err != nil && isAdmin == false {
+		log.Error("get tenant id failed")
+		err = ErrInternalError
+		return err
```

Comment:
Hi @sunfch  What does error represent, grpc framework error or business error? And what's the different between out.ErrorCode and err?

Reply:
@sfzeng ,  I return error always when error happened. If you want to separate framework and business error, I will modify it to return nil when business error happened